### DISCUSSION TITLE
how can use get_option() in project()

### DISCUSSION THREAD

**Discussion by Freed-Wu** [mesonbuild/meson#13141] [Q&A]
```meson
project(
  'test',
  ['c'],
  version: get_option('version'),
)
```

Cannot work, it said `version` is unknown.

**Comment by eli-schwartz** [comment]
Why would "version" ever be a choice that the user building the project is making?

**Reply by Freed-Wu** [reply]
There just is an example, for other things (langauge, ...) they also cannot use `get_option()`

And for version, I figure out one situation:

In github workflow, there is an enviroment variable named `$GITHUB_REF_NAME` contains `git tag` information. We can use `meson setup build -Dversion=$GITHUB_REF_NAME` to pass version to `meson`, to make sure we can build finial project with the version determined from `git tag`.

**Reply by eli-schwartz** [reply]
But what is the use case?

Version makes no sense -- using `meson setup build -Dversion=$GITHUB_REF_NAME` in GitHub actions means you can't build the project *outside* of GitHub actions with working project information.

Language isn't a good use case either -- if you have conditionally used languages, please use add_languages with the `required` kwarg at the point where you are *using* other languages (maybe behind an OS conditional, maybe behind a feature option).

**Reply by Freed-Wu** [reply]
> you can't build the project outside of GitHub actions with working project information.

out of github action, we can `meson setup build` use default value of `meson.options`.

> Language isn't a good use case either

Just an example :smile: `required` should be better choice for this situtation.

**Reply by eli-schwartz** [reply]
Just don't do broken stuff that depends on GitHub actions?

If using a default value in meson.options is okay then simply put that same value in meson.build instead, and don't bother with GITHUB_REF_NAME no matter what.

If what you actually want is to determine the version based on the state of the git repo, you should be using `git describe` anyway.
